### PR TITLE
563/fix eth sign v4 signature not supported

### LINKED ISSUE #563: `eth_signTypedData_v4` not supported
**Describe the bug**
Error signing the transaction

**Details**
`eth_signTypedData_v4` not supported.

**Additional context**
This comes from the feedback button, but no email was left to get more context on how to reproduce.

The message the user gets is:
`the method eth_signTypedData_v4 does not exist/is not available`

There's 6 methods created for historic reasons, maybe we can fallback to v3, and eth_sign as last resources?
* eth_sign
* personal_sign
* signTypedData (currently identical to signTypedData_v1)
* signTypedData_v1
* signTypedData_v3
* signTypedData_v4

It would be great to get to reproduce this.

More info in the device:
image

I bet is a webview within an app, probably imToken if I have to bet on one (seeing the user is from Singapore)

@nlordell WDYT?

### PR DESCRIPTION
**Pull request by alfetopito**
# Update 2021-05-27

- Using new version of contracts NPM lib
- No longer using my hacked version of ethers.js lib
- This change fixes 2 wallets:
    - TrustWallet mobile browser
    - Opera browser wallet (Fantom)
- Trying different signing methods in sequence:
    - v4 (default and preferred)
    - v3 (when v4 not available)
    - eth_sign (when v3 not available)

# Summary

Closes #563 

After this change, I'm able to sign orders on Trust wallet dapp browser.
Need to test with more wallets of this type (injected)

# Testing
1. Using Trust wallet mobile dapp browser, open this PR's deployed app
2. Place an order like you normally do (fill in all fields, allow a token if needed)
- [ ] A signature request of a typed data should be requested
- [ ] The signature should succeed and the order should be placed successfully

3. Using Opera browser, setup the built in crypto wallet
4. In your mobile device, install Opera and setup crypto wallet there as well
4. Open this PR's deployed app
5. Place an order
- [ ] On your phone, you should see the signature request of the hash
- [ ] The signature should succeed and the order should be placed successfully

### CODE REVIEW

**Comment by alfetopito** [comment]
This change alone won't do it.
We need also a change to `ethers.js`, as Trustwallet needs to use `personal_sign` instead of `eth_sign`, like metamask.
Opened a PR against their repo https://github.com/ethers-io/ethers.js/pull/1542

**Comment by anxolin** [comment]
Maybe @MareenG could also run some tests in some of the failing wallets https://docs.google.com/spreadsheets/u/1/d/177RoN3x0gHDwKAgu-TXUYEuz-z-dPu3RybZvOumRrAA/edit#gid=0

File: cypress/support/commands.js [hunk 1]
```diff
@@ -4,7 +4,7 @@
 // https://on.cypress.io/custom-commands
 // ***********************************************
 
-import { JsonRpcProvider } from '@ethersproject/providers'
+import { JsonRpcProvider } from 'ethers_providers_alfetopito'
```
Comment (anxolin) [inline cypress/support/commands.js:7]: Cool, let's add an issue also so we don't forget to bring this back to etheres.
Awesome job patching them

File: src/custom/utils/trade.ts [hunk 2]
```diff
@@ -10,6 +10,7 @@ import { EcdsaSignature, Signature } from '@gnosis.pm/gp-v2-contracts'
 
 const DEFAULT_SIGNING_SCHEME = SigningScheme.EIP712
 const METAMASK_SIGNATURE_ERROR_CODE = -32603
+const V4_ERROR_MSG_REGEX = /eth_signTypedData_v4 does not exist/i
```
Comment (anxolin) [inline src/custom/utils/trade.ts:13]: Do we need a regex for this? it feels like equality on the message would do, right?

**Comment by MareenG** [comment]
Yes, I will do some more tests. Are there some you would consider as most important to test?

File: src/custom/utils/trade.ts (hunk 2 above)
Comment (anxolin) [reply src/custom/utils/trade.ts:13]: okoko,
u know i love regex, all good. Just for trolling a bit https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith

**Comment by anxolin** [comment]
For IOS people! now you can re-enable the Dapp browser that was removed in a past version https://link.trustwallet.com/browser_enable

**Review by anxolin — COMMENTED** [review]
It didn't work for me 😅

image

**Comment by alfetopito** [comment]
> @alfetopito is the new ethers package being used here or still yours? is there anything pending in this PR or should i review?

@david Yes, it should be mine :)
Ethers.js hasn't released a new version yet. Neither has my PR been merged.

How to review? 
Not much in terms of code change, just one additional conditional on the `trade` utils module.
I'd say the main things is to test the app, which is basically test using TrustWallet's dApp browser (see `Test steps`)
The test must be done on mainnet, and you can find details for a shared wallet on 1password to setup it up in your phone.

@anxolin Can you try again? Several times?
Their RPC connection is not very reliable. Doesn't work all time for me.

**Comment by anxolin** [comment]
> @anxolin Can you try again? Several times?
> Their RPC connection is not very reliable. Doesn't work all time for me.

I've tried like 20 times, it looks stable to me the read operations, even wrap operations. It's just the signing operation is not working

**Comment by alfetopito** [comment]
> > @anxolin Can you try again? Several times?
> > Their RPC connection is not very reliable. Doesn't work all time for me.
> 
> I've tried like 20 times, it looks stable to me the read operations, even wrap operations. It's just the signing operation is not working
> 
> It's the same error I get in cowswap.exchange by the way

:disappointed: 

Can't help you there. You'll have to debug this one for me, since I have no apple device to test with.

**Comment by W3stside** [comment]
> @david Yes, it should be mine :)

lmaoooo i love what you tagged me as 😂 😂 😂

**Comment by alfetopito** [comment]
> > @david Yes, it should be mine :)
> 
> lmaoooo i love what you tagged me as joy joy joy

I got that from @anxolin always tagging the wrong Michel =P

But anyway, it's not too far off. He's your neighbor ;)

**Comment by anxolin** [comment]
I tried to follow trust debug instructions, but they are hard to follow for IOS.

I opened this issue:
https://github.com/trustwallet/trust-web3-provider/issues/148

Probably we will need to inject some remote debug library for this. According to what they say, they implemented v3, which we also default back to.

**Comment by alfetopito** [comment]
> I tried to follow trust debug instructions, but they are hard to follow for IOS.
> 
> I opened this issue:
> [trustwallet/trust-web3-provider#148](https://github.com/trustwallet/trust-web3-provider/issues/148)
> 
> Probably we will need to inject some remote debug library for this. According to what they say, they implemented v3, which we also default back to.

Where do you see we default to `v3`?

We currently don't do that, AFAIAA

**Comment by alfetopito** [comment]
I know it's broken.
Needs the changes from https://github.com/gnosis/gp-v2-contracts/pull/650 to be released first.

**Comment by alfetopito** [comment]
> Trust wallet is addressing the issue that we have in IOS: [trustwallet/trust-web3-provider#149](https://github.com/trustwallet/trust-web3-provider/pull/149)

Great news!

Although, I've made such a nice logic for trying different versions sequentially :disappointed: 

No matter, I'm sure it'll be useful for other wallets.

**Comment by alfetopito** [comment]
Currently broken.
Depends on https://github.com/gnosis/gp-v2-contracts/pull/659 be published to NPM

**Review by anxolin — APPROVED** [review]
